build/patches/experimental-user-scripts-support.patch

```diff
@@ -0,0 +1,10299 @@
++    // The host may be empty (e.g., in the case of *://*/* - in that case, only
++    // append '*' instead of '*.'.
++    subdomains = host.empty() ? "*" : "*.";
++  }
++
++  base::StringPiece scheme_separator =
++      IsStandardScheme(scheme) ? url::kStandardSchemeSeparator : ":";
++
++  std::string pattern_str = base::StrCat(
++      {scheme, scheme_separator, subdomains, host, ":", port, path});
++
++  URLPattern pattern(intersection_schemes);
++  ParseResult result = pattern.Parse(pattern_str);
++  // TODO(devlin): I don't think there's any way this should ever fail, but
++  // use a CHECK() to flush any cases out. If nothing crops up, downgrade this
++  // to a DCHECK in M72.
++  CHECK_EQ(ParseResult::kSuccess, result);
++
++  return pattern;
++}
++
++bool URLPattern::MatchesAnyScheme(
++    const std::vector<std::string>& schemes) const {
++  for (auto i = schemes.cbegin(); i != schemes.cend(); ++i) {
++    if (MatchesScheme(*i))
++      return true;
++  }
++
++  return false;
++}
++
++bool URLPattern::MatchesAllSchemes(
++    const std::vector<std::string>& schemes) const {
++  for (auto i = schemes.cbegin(); i != schemes.cend(); ++i) {
++    if (!MatchesScheme(*i))
++      return false;
++  }
++
++  return true;
++}
++
++bool URLPattern::MatchesSecurityOriginHelper(const GURL& test) const {
++  // Ignore hostname if scheme is file://.
++  if (scheme_ != url::kFileScheme && !MatchesHost(test))
++    return false;
++
++  if (!MatchesPortPattern(base::NumberToString(test.EffectiveIntPort())))
++    return false;
++
++  return true;
++}
++
++bool URLPattern::MatchesPortPattern(base::StringPiece port) const {
++  return port_ == "*" || port_ == port;
++}
++
++std::vector<std::string> URLPattern::GetExplicitSchemes() const {
++  std::vector<std::string> result;
++
++  if (scheme_ != "*" && !match_all_urls_ && IsValidScheme(scheme_)) {
++    result.push_back(scheme_);
++    return result;
++  }
++
++  for (size_t i = 0; i < base::size(kValidSchemes); ++i) {
++    if (MatchesScheme(kValidSchemes[i])) {
++      result.push_back(kValidSchemes[i]);
++    }
++  }
++
++  return result;
++}
++
++std::vector<URLPattern> URLPattern::ConvertToExplicitSchemes() const {
++  std::vector<std::string> explicit_schemes = GetExplicitSchemes();
++  std::vector<URLPattern> result;
++
++  for (std::vector<std::string>::const_iterator i = explicit_schemes.begin();
++       i != explicit_schemes.end(); ++i) {
++    URLPattern temp = *this;
++    temp.SetScheme(*i);
++    temp.SetMatchAllURLs(false);
++    result.push_back(temp);
++  }
++
++  return result;
++}
++
++// static
++const char* URLPattern::GetParseResultString(
++    URLPattern::ParseResult parse_result) {
++  return kParseResultMessages[static_cast<int>(parse_result)];
++}
+diff --git a/components/user_scripts/common/url_pattern.h b/components/user_scripts/common/url_pattern.h
+new file mode 100755
+--- /dev/null
++++ b/components/user_scripts/common/url_pattern.h
```

Comment:
Where did you find these? They have been removed at some point?

Reply:
they are in extensions, here
https://source.chromium.org/chromium/chromium/src/+/master:extensions/common/url_pattern.h

Reply:
They cannot be included directly? Do we have changes to these?

Reply:
> They cannot be included directly?

no, better not

> Do we have changes to these?

yes, look here
https://github.com/bromite/bromite/compare/master...uazo:enable-user-scripts-src
you better use a git client to see changes

Reply:
I see; I will think of some automation after we merge this patch so that we keep them in sync.

Reply:
yes, for now i am monitoring the changes with wachete
something has changed in the last 6 months, but nothing much, except the transformation from IPC to mojo for calling the userscripts.
that's something I'll have to check